zcash_primitives/src/sapling/address.rs

```diff
@@ -0,0 +1,118 @@
+use group::{Group, GroupEncoding};
+
+use super::{
+    keys::Diversifier,
+    note::{Note, Rseed},
+};
+
+/// A Sapling payment address.
+///
+/// # Invariants
+///
+/// `pk_d` is guaranteed to be prime-order (i.e. in the prime-order subgroup of Jubjub,
+/// and not the identity).
+#[derive(Clone, Debug)]
+pub struct PaymentAddress {
+    pk_d: jubjub::SubgroupPoint,
+    diversifier: Diversifier,
+}
+
+impl PartialEq for PaymentAddress {
+    fn eq(&self, other: &Self) -> bool {
+        self.pk_d == other.pk_d && self.diversifier == other.diversifier
+    }
+}
+
+impl Eq for PaymentAddress {}
+
+impl PaymentAddress {
+    /// Constructs a PaymentAddress from a diversifier and a Jubjub point.
+    ///
+    /// Returns None if `pk_d` is the identity.
+    pub fn from_parts(diversifier: Diversifier, pk_d: jubjub::SubgroupPoint) -> Option<Self> {
+        if pk_d.is_identity().into() {
+            None
+        } else {
```

Comment:
```suggestion
        } else {
            // Check that the diversifier is valid
            diversifier.g_d()?;
```
and add a notice of the incompatible change to the release notes.

(The protocol spec omits to specify in [section 5.6.3.1](https://zips.z.cash/protocol/protocol.pdf#saplingpaymentaddrencoding) that `\mathsf{DiversifyHash^{Sapling}}(d)` MUST NOT return `\bot`. It was intended to specify that, and there should be no incompatibility from also enforcing it here, since it was already enforced by `PaymentAddress::from_bytes`. I have filed https://github.com/zcash/zips/issues/659 for the spec omission.)

Reply:
Per https://github.com/zcash/librustzcash/pull/753#discussion_r1064859474, this PR is intentionally not changing the existing public API. I will be doing this in a subsequent PR.

Reply:
👍